File: 47.全排列-ii.py
#
# @lc app=leetcode.cn id=47 lang=python3
#
# [47] 全排列 II
#

# @lc code=start
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        current = 0
        res = self.backtrack(nums, current)
        return res

    def backtrack(self, nums: List[int], current: int) -> List[List[int]]:
        tmp=nums.copy()
        res = []
        if current == len(tmp):
            res.append(tmp)
            return res
        current += 1
        res.extend(self.backtrack(tmp, current))
        current -= 1
        for i in range(current):
            if tmp[i] == tmp[current]:
                break
            tmp[i], tmp[current] = tmp[current], tmp[i]
            current += 1
            res.extend(self.backtrack(tmp, current))
            current -= 1
            tmp[i], tmp[current] = tmp[current], nums[i]
        return res
# @lc code=end

File: 47.全排列-ii.cpp
/*
 * @lc app=leetcode.cn id=47 lang=cpp
 *
 * [47] 全排列 II
 */

// @lc code=start
class Solution
{
public:
    vector<vector<int>> permuteUnique(vector<int> &nums)
    {
        sort(nums.begin(), nums.end());
        int current_len = 0;
        vector<vector<int>> res;
        function<void(void)> backtrack = [&](void) {
            if (current_len == nums.size())
            {
                res.push_back(nums);
                return;
            }
            current_len++;
            backtrack();
            current_len--;
            for (int i = 0; i < current_len; i++)
            {
                if (nums[i] == nums[current_len])
                    break;
                swap(nums[i], nums[current_len]);
                current_len++;
                backtrack();
                current_len--;
                swap(nums[i], nums[current_len]);
            }
        };
        backtrack();
        return res;
    }
};
// @lc code=end
